### DISCUSSION TITLE
version 4.2.0

### DISCUSSION THREAD

**Discussion by hvianna** [hvianna/audioMotion-analyzer#56] [Announcements]
### Added:

+ new `stopTracks` argument to [`disconnectInput()`](README.md#disconnectinput-node-stoptracks-) - thanks **@hookedupjoe** ([#52](https://github.com/hvianna/audioMotion-analyzer/issues/52));
+ [`destroy()`](README.md#destroy) - thanks **@lucienimmink** for the [suggestion](https://github.com/hvianna/audioMotion-analyzer/issues/53);
+ [`maxFPS`](README.md#maxfps-number) - set the maximum desired animation frame rate;
+ [`peakLine`](README.md#peakline-boolean) - show peaks as a continuous line for **Graph** [`mode`](README.md#mode-number);
+ [`start()`](README.md#start) and [`stop()`](README.md#stop) methods.

### Fixed:

+ Corrected ESM exports in `package.json` - thanks **@cprussin** ([#54](https://github.com/hvianna/audioMotion-analyzer/pull/54)).

### Improved:

+ Peak hold and decay times and radial spin speed should now remain consistent on any frame rate;
+ [`showPeaks`](README.md#showpeaks-boolean) now works for **Graph** mode in **radial** view;
+ Code quality and documentation improvements - also thanks **@Staijn1** for [#55](https://github.com/hvianna/audioMotion-analyzer/pull/55).

<hr /><em>This discussion was created from the release <a href='https://github.com/hvianna/audioMotion-analyzer/releases/tag/4.2.0'>version 4.2.0</a>.</em>

**Comment by TF3RDL** [comment]
Great update. BTW, I still have ideas to add in upcoming releases; a 1/48th octave and Lanczos FFT bin interpolation mode would be cool like this:
![48th octave notes spectrum](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/d9fb0c3f-26ef-4b3e-a199-ec05efa4e8c2)

Also, I like to see the color gradient (both light and dark mode) used by foobar2000's built-in spectrum like this:
![mel frequency bargraph spectrum](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/dfab6370-c293-40a2-b481-e7c260abd185)

And the ability to set frequency label display to decades just like in [the mockup of obviously-nonexistent foo_fft_analyzer](https://codepen.io/TF3RDL/pen/poQJwRW) that would likely to be used as a mean of what [visualization support](https://github.com/TheQwertiest/foo_spider_monkey_panel/discussions/222) looks like for SMP component (foo_spider_monkey_panel) for foobar2000

BTW, I think it is possible to untie `smoothingTimeConstant` decay time from FPS in a similar way to peak decay times, except it involves exponential functions and a making untie peak decay/smoothing times from FPS an option rather than being always on

**Reply by hvianna** [reply]
I appreciate the feedback and ideas, as always! That Lanczos interpolation looks so smooth!
Right now I'm working on a much-needed update to audioMotion (the player) and then I'll be back to improving the analyzer.
Keep them coming! :)

**Reply by TF3RDL** [reply]
BTW, the Lanczos interpolation works best when the FFT bin interpolation operates in linear amplitude scale, which is inherent to custom FFT (and `getFloatFrequencyData()` are in decibels scale) that I've used on my Web Audio API visualizations to support custom window functions and non-power of 2 FFT sizes (implementing this makes it possible to set FFT size in milliseconds as opposed to samples, which change as I noticed that time/frequency resolution in ms/Hz changes when I apply resampling DSPs on foobar2000)

**Reply by JesusGod-Pope666-Info** [reply]
Awesome I hope you make it easier for reaching out for files on the server :) That would be awesome.

**Reply by TF3RDL** [reply]
As finally I've get the grasp of OffscreenCanvas (yes, you can use it without Web Workers) when implementing a spectrogram mode for [my spectrum analyzer project](https://codepen.io/TF3RDL/pen/VwENJbB), a spectrogram view (#38) would be nice, and it can be both spectrum and spectrogram at the same time like this:
![spectrum and spectrogram](https://github.com/hvianna/audioMotion-analyzer/assets/102759506/e5b1b54a-ae87-4809-8acb-ab47c7d05600)

**Reply by hvianna** [reply]
Oh this is a good layout idea for having both the spectrum and the spectrogram simultaneously! Also, I need to take a deeper look into OffscreenCanvas to see if I can use it to improve rendering performance.

**Comment by nikolaict** [comment]
Hello,

Is there any possibility to implement audioMotion-analyzer for ES5 version?

Thank you

**Reply by hvianna** [reply]
Hello!
Well, it should be possible of course, but ES6 has a lot of features that improve code quality and performance.
Can you elaborate on this so I can better understand what your needs are? Is it the `import` statement that's causing you any trouble?

**Reply by nikolaict** [reply]
Of course, but the project is developed in ES5, so I need a syntax file with ES5 code.

In addition to import and export class.

Thank you

**Reply by Staijn1** [reply]
Hi Nikolai,

If I understand you correctly you would like the audiomotion-analyzer in ES5 format.
I would assume because your own project uses ES5 aswell. 

I'm not very familiar with this, but wouldn't it be possible to import Audiomotion and let some build tool convert it to ES5, like babel or webpack?

@hvianna I will admit I have had the thought to port Audiomotion-Analyzer to Typescript since the types already exist which is the biggest pain. When you have typescript you can set up the target outputs, like ES6 or ES5. The request above could be a driver to change to Typescript. What do you think?

**Reply by hvianna** [reply]
@nikolaict I'll assume you're not using a bundler, like Webpack, or similar. Have you tried setting your `script` element to `type="module"` so it accepts the `import` statement? (note you **can't** use`type="text/javascript"`) 

```
<script src="index.js" type="module">
```
Other than that, it should have no impact in your ES5 code.

@Staijn1 Theoretically you can use [Babel to transpile ES6 features to ES5](https://www.tutorialspoint.com/babeljs/babeljs_transpile_es6_features_to_es5.htm). I'm still not sure there's really a need for this though.
